libcontainer/container_linux.go

```diff
@@ -454,24 +457,42 @@ func (c *Container) newParentProcess(p *Process) (parentProcess, error) {
 	}
 	logFilePair := filePair{parentLogPipe, childLogPipe}
 
-	cmd := c.commandTemplate(p, childInitPipe, childLogPipe)
-	if !p.Init {
-		return c.newSetnsProcess(p, cmd, messageSockPair, logFilePair)
-	}
-
-	// We only set up fifoFd if we're not doing a `runc exec`. The historic
-	// reason for this is that previously we would pass a dirfd that allowed
-	// for container rootfs escape (and not doing it in `runc exec` avoided
-	// that problem), but we no longer do that. However, there's no need to do
-	// this for `runc exec` so we just keep it this way to be safe.
-	if err := c.includeExecFifo(cmd); err != nil {
-		return nil, fmt.Errorf("unable to setup exec fifo: %w", err)
+	// Make sure we use a new safe copy of /proc/self/exe or the runc-dmz
+	// binary each time this is called, to make sure that if a container
+	// manages to overwrite the file it cannot affect other containers on the
+	// system. For runc, this code will only ever be called once, but
+	// libcontainer users might call this more than once.
+	p.closeClonedExes()
+	var (
+		exePath string
+		// only one of dmzExe or safeExe are used at a time
+		dmzExe, safeExe *os.File
+	)
+	if dmz.IsSelfExeCloned() {
+		// /proc/self/exe is already a cloned binary -- no need to do anything
+		logrus.Debugf("skipping binary cloning -- /proc/self/exe is already cloned!")
+		exePath = "/proc/self/exe"
+	} else {
+		dmzExe, err = dmz.Binary(c.root)
+		if err == nil {
+			// We can use our own executable without cloning if we are using
+			// runc-dmz.
+			exePath = "/proc/self/exe"
+			p.clonedExes = append(p.clonedExes, dmzExe)
+		} else if errors.Is(err, dmz.ErrNoDmzBinary) {
+			logrus.Debugf("runc-dmz binary not embedded in runc binary, falling back to /proc/self/exe clone")
+			safeExe, err = dmz.CloneSelfExe(c.root)
+			if err != nil {
+				return nil, fmt.Errorf("unable to create safe /proc/self/exe clone for runc init: %w", err)
+			}
+			exePath = "/proc/self/fd/" + strconv.Itoa(int(safeExe.Fd()))
+			p.clonedExes = append(p.clonedExes, safeExe)
+		} else {
+			return nil, fmt.Errorf("failed to create runc-dmz binary clone: %w", err)
+		}
```

Comment:
Can we have an env var to forcibly disable DMZ, or fallback to non-DMZ on a failure to execute DMZ?

Because DMZ might be potentially incompatible with some LSM (with distro's default profile).

Reply:
In order to detect if an LSM would block the final exec (after having the container's LSM labels applied), we would need to do the test then but at that point we cannot do a memfd clone of `/proc/self/exe`.

If we did a test-exec here, a failure would mean that `/proc/self/exe` would also presumably fail, at which point we are going to error out no matter what (and making both be option by accident because they use the same cloning code would re-introduce the CVE).

We could do an environment variable though.

Reply:
>  I think you can't set annotations with upstream Docker

You can:
- https://github.com/docker/cli/pull/4156

Reply:
Ah right. But I'm sure you'd agree setting it with annotations is pretty ugly. :sweat_smile:

Reply:
ugly, but acceptable as a workaround

Reply:
Can you set annotations with Kubernetes as well?

Reply:
I went with an environment variable -- `RUNC_DMZ=legacy`.